From 7e7ac84164d0ece6d77c448ab4bd0f3c781878ea Mon Sep 17 00:00:00 2001
From: Jens Ole Lauridsen <jlauridsen@google.com>
Date: Tue, 30 Jun 2015 14:11:36 -0700
Subject: [PATCH] Always change the minimize state of a LightToolWindow on anchor changes.

When an anchor position (left/right) is changed on a visible
LightToolWindow, the minimize button may up on the wrong side.
Also fix the hide icon to point to the correct direction after the
anchor change.

Change-Id: Ib6a22fff480040ec81e61994604c50d9f1201f92
---

diff --git a/platform/platform-impl/src/com/intellij/designer/LightToolWindow.java b/platform/platform-impl/src/com/intellij/designer/LightToolWindow.java
index ce3cb33..80e2c5a 100644
--- a/platform/platform-impl/src/com/intellij/designer/LightToolWindow.java
+++ b/platform/platform-impl/src/com/intellij/designer/LightToolWindow.java
@@ -65,6 +65,7 @@
   private final String myWidthKey;
   private final JPanel myMinimizeComponent;
   private final AnchoredButton myMinimizeButton;
+  private final JPanel myActionPanel;
 
   private final TogglePinnedModeAction myToggleAutoHideModeAction = new TogglePinnedModeAction();
   private final ToggleDockModeAction myToggleDockModeAction = new ToggleDockModeAction();
@@ -116,21 +117,21 @@
     titleLabel.setFont(UIUtil.getLabelFont(UIUtil.FontSize.SMALL));
     header.add(titleLabel, BorderLayout.CENTER);
 
-    JPanel actionPanel = new JPanel(new FlowLayout(FlowLayout.CENTER, 5, 0));
-    actionPanel.setBorder(IdeBorderFactory.createEmptyBorder(3, 0, 2, 0));
-    actionPanel.setOpaque(false);
-    header.add(actionPanel, BorderLayout.EAST);
+    myActionPanel = new JPanel(new FlowLayout(FlowLayout.CENTER, 5, 0));
+    myActionPanel.setBorder(IdeBorderFactory.createEmptyBorder(3, 0, 2, 0));
+    myActionPanel.setOpaque(false);
+    header.add(myActionPanel, BorderLayout.EAST);
 
     if (actions != null) {
       for (AnAction action : actions) {
-        addAction(actionPanel, action);
+        addAction(myActionPanel, action);
       }
 
-      actionPanel.add(new JLabel(AllIcons.General.Divider));
+      myActionPanel.add(new JLabel(AllIcons.General.Divider));
     }
 
-    addAction(actionPanel, new GearAction());
-    addAction(actionPanel, new HideAction());
+    addAction(myActionPanel, new GearAction());
+    addAction(myActionPanel, new HideAction());
 
     JPanel contentWrapper = new JPanel(new BorderLayout());
     contentWrapper.setBorder(IdeBorderFactory.createBorder(SideBorder.TOP));
@@ -225,16 +226,15 @@
         setContentComponent(null);
       }
     }
-    else {
-      String key = getMinKey();
-      if (minimizeParent.getClientProperty(key) == myMinimizeComponent) {
-        minimizeParent.putClientProperty(key, null);
-      }
-      minimizeParent.putClientProperty(isLeft() ? RIGHT_MIN_KEY : LEFT_MIN_KEY, myMinimizeComponent);
-      minimizeParent.revalidate();
+    String key = getMinKey();
+    if (minimizeParent.getClientProperty(key) == myMinimizeComponent) {
+      minimizeParent.putClientProperty(key, null);
     }
+    minimizeParent.putClientProperty(isLeft() ? RIGHT_MIN_KEY : LEFT_MIN_KEY, myMinimizeComponent);
+    minimizeParent.revalidate();
 
     myAnchor = newAnchor;
+    updateHideIcons();
     configureBorder();
     updateWidth();
 
@@ -245,6 +245,11 @@
     minimizeParent.putClientProperty(IGNORE_WIDTH_KEY, null);
   }
 
+  private void updateHideIcons() {
+    myActionPanel.remove(myActionPanel.getComponentCount() - 1);
+    addAction(myActionPanel, new HideAction());
+  }
+
   private void updateContent(boolean show, boolean flag) {
     myShowContent = show;
 
